### PR TITLE
Initial workflow runtime

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

Contains the initial workflow runtime with no features. Features will be added in the next branch after this. This is only open early for reviewers. Most suggestions will be added in the next branch.

### CODE REVIEW

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowInstance.cs:194]: Clarify in comment that we intend to be in order of: patch, signal, non-query, query

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowInstance.cs:482]:
```suggestion
                    ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
```
Properly keep stack trace.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowWorker.cs:22]: I set this to 2 in Python because it's just a slow language, but can probably set to 1 in .NET like Go/Java. Java technically allows this to be configured, but other SDKs don't and I like "on or off".

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowWorker.cs:28]: Document somewhere how we're handling deadlocked workflows.

What we're doing here to prevent deadlocked workflows from eating up more thread pool space is failing the WFT when deadlocked, and then adding here. So we still proceed with cache eviction, but any future WFTs are responded to immediately with a deadlock without even recreating the workflow. This entry is removed once the task finally completes (if ever).

@Sushisource, @bergundy suggested that instead of this approach (evicting and just eagerly failing new activations), I should just never send back completion for eviction activation while deadlocked. Think that's a better approach? It'll properly consume a cache task slot right? And is there any harm to core in an eviction not being handled immediately but hanging (possibly forever)? I assume shutdown will not send a poll shutdown error ending polling until all outstanding remove-from-cache's are handled?

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowWorker.cs:51]: We don't technically need this instance factory concept, but I am trying to keep the serializable/sandbox separation alive internally in .NET even though we'll never use a sandbox most likely.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowWorker.cs:90]: This isn't right, the `ContinueWith` task is different than the one we're removing. We need to store the original task in the dictionary, _then_ add the continue with.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [inline src/Temporalio/Worker/WorkflowWorker.cs:146]: Create this with a cancel token that we cancel when it didn't timeout. Otherwise this task will complete later unnecessarily (see https://devblogs.microsoft.com/oldnewthing/20220505-00/?p=106585).

I can't find a much cheaper way to timeout a task than create a delayed task for every single task run :-( https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync is only supported in newer .NET versions. Looking [at the source](https://github.com/dotnet/runtime/blob/dfb3d24cec0a227817420de0ae95f45b210e7c25/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2798-L2816) it basically does what Delay does (timer queue timer) though maybe a couple fewer steps. I will only move to it if microbenchmarks confirm necessity.

I should at least at the above in a comment here.

File: tests/Temporalio.Tests/Worker/WorkflowWorkerTests.cs
Comment (cretz) [inline tests/Temporalio.Tests/Worker/WorkflowWorkerTests.cs:131]: This is what I consider the full set of tests to confirm the workflow runtime is complete for this SDK. Feel free to suggest any I have missed (or ask for clarification on any if needed).

**Review by cretz — COMMENTED** [review]
Ignore failing CI, it is due to https://github.com/dotnet/format/issues/1800 and I don't want to disable code style in build yet.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (Sushisource) [inline src/Temporalio/Worker/WorkflowInstance.cs:139]: Probably worth at least logging something in that instance

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (Sushisource) [inline src/Temporalio/Worker/WorkflowInstance.cs:261]: You shouldn't have to do this - why don't you open an issue in core and I can handle this there and then we can get rid of this code.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (Sushisource) [inline src/Temporalio/Worker/WorkflowInstance.cs:362]: Seems like this state could be made unrepresentable by having completion be passed through the relevant functions instead of as a class field?

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (Sushisource) [reply src/Temporalio/Worker/WorkflowWorker.cs:28]: You're allowed to not complete a task forever, and the slot will be occupied. Shutdown will never resolve until all outstanding work is replied to.

**Review by Sushisource — COMMENTED** [review]
Looks like it's making sense to me :+1:

File: src/Temporalio/Worker/TemporalWorker.cs
Comment (bergundy) [inline src/Temporalio/Worker/TemporalWorker.cs:78]: Hmm.. might be cleaner to do this in the workflow worker instead of the the top level one.

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (bergundy) [inline src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: nit: there's not really a concept of remote activities, we've used it before but I think it'd be a bit clearer to use NonLocal instead.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:28]: @Sushisource - do you think that's a better approach than what I have today where I do evict but still kinda occupy a slot locally? I'm thinking so because I bet that task I'm holding has a reference to my task scheduler which means the whole workflow is probably still in memory.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:139]: I think we decided this is just an accepted Temporal feature in all SDKs. Not even a warn-based one.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:261]: Opened https://github.com/temporalio/sdk-core/issues/497

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:362]: Correct for this PR, but the next one will have things on outbound interceptor (e.g. starting a timer) that is too far separated from this to have a completion passed through.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: Would be nice to parallelize this

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: Can we have some test that verifies every payload goes through these methods?
In TS we can catch this at compile time but the type system here obviously isn't complex enough for that.
I mostly worry about forgetting to handle new jobs and commands as we add those.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: What about headers? those should be decoded / encoded here.
Mostly for skipping the deadline detector in workflow.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:195]: Core should just do this for us, no need to repeat this in every lang.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:400]: This condition should be tested (when implemented)

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (Sushisource) [reply src/Temporalio/Worker/WorkflowWorker.cs:28]: I think that'd be swell - but you do still want to fail the activation so that the user ends up seeing WFT failure "because deadlocked" in their history. As long as that still happens the approach you describe makes sense to me.

File: src/Temporalio/Worker/WorkflowInstanceDetails.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstanceDetails.cs:29]: You don't need to save these, better to free up memory used by the payloads.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowWorker.cs:28]: > I assume shutdown will not send a poll shutdown error ending polling until all outstanding remove-from-cache's are handled?

That's the desired behavior but I'm not convinced that's what currently happens. We'll need to address that core side.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowWorker.cs:122]: @Sushisource can / does core ensure that remove from cache would always be last?

Comment (bergundy) [reply src/Temporalio/Worker/WorkflowWorker.cs:122]: When would lang get remove from cache + other jobs?

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowWorker.cs:194]: Hmm... looks like you might have both Failure and Successful set in the completion here.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowWorker.cs:217]: This is redundant / inaccurate

File: src/Temporalio/Worker/TemporalWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/TemporalWorker.cs:78]: The workflow worker is not disposable and I want this registered/unregistered on construct/dispose, not start/shutdown. No particular reason, just feels better to do it on these boundaries.

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (cretz) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: Hrmm, this is what it's called in Python. I assumed "remote" means "non-local". You're saying this should be called `DisableNonLocalActivities`? Maybe I should change to Go and Java wording of `LocalActivityWorkerOnly`?

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: I am not convinced there is a ton of value in parallelizing. In fact, I think there's more value in batching. Shall we change our codecs to submit these things in batches? Or do we accept code complication is not enough benefit to contort the code to try to do multiple of these things at once?

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: Yes. I was thinking about writing a custom codec test that recursively, reflectively iterates over every payload in the nested proto hierarchy (excluding search attributes), and creates a new object all the way from root (i.e. clones) and sets the payload and runs root through this and confirms the payload was converted. This should catch new payloads. We could also write a code generator like the Go proxy does, but I think this is fine for now.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:195]: Feel free to create issues in core when you come across these things. I won't let it slow down my progress.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: Hrmm, I think I missed headers in Python (and the aforementioned reflection test would catch these for .NET). Will fix here, but what should I do about Python? Leave headers unencoded for compat reasons?

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:400]: Which "condition"? General workflow cancellation in its multiple bubbled forms? Yes, that will be part of tests.

File: src/Temporalio/Worker/WorkflowInstanceDetails.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstanceDetails.cs:29]: True. I will keep them here but can avoid setting the whole details object as a field on the instance

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:194]: Protobuf `oneof`s properly handle mutual exclusivity internally

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:217]: Yes, will remove, bad copy/pasta

**Comment by cretz** [comment]
@bergundy and @Sushisource - I plan on making required changes in next PR, see any issues with that? I can either merge this when that PR comes about or I can close this and have that PR be all this new stuff + a bunch of new things? Preference?

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (bergundy) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: I think `DisableNonLocalActivities` or better yet `EnableNonLocalActivities` is the clearest.
`LocalActivityWorkerOnly` isn't technically correct since the worker runs workflows (which is why Spencer renamed it in Core).

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: Value is shorter task processing times and reduced chance of workflow task timeouts especially for remote codecs.
Not blocking the PR, we don't parallelize all invocations in other SDKs.

Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: Just thought it would be nice to consider since this is a fresh implementation.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: Agree but could use an issue.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: Headers should be encoded, they may contain sensitive information.

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:195]: 👍

**Review by bergundy — APPROVED** [review]
Any unaddressed comments can be addressed in a follow up PR.
Approving to maintain the momentum.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (Sushisource) [reply src/Temporalio/Worker/WorkflowWorker.cs:122]: It's always last in the job list, but you can get other jobs if the WFT is about to be replied to and the cache is at the limit. So we finish the task, reply, then evict to make room for someone else

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (cretz) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: We would prefer false to be the default IMO (like Go, proto, unset .NET boolean primitive, etc). So, what should we name it in the interest of clarity and consistency and default-false semantics? `DisableNonLocalActivities` is an inconsistent, ugly double negative.

In general, can you provide an adjective for "non-local" that Temporal can use company wide? Why can "remote" not be it since it's the opposite of local?

> LocalActivityWorkerOnly isn't technically correct since the worker runs workflows

It's only incorrect depending on how you read it. You read "local-activity worker only" and I read it as "local activity-worker only" completely unrelated to workflow workers. In fact, most of our worker options apply to either activity workers or workflow workers, not both. I mean I guess I can rename `GracefulShutdownTimeout` to `GracefulShutdownActivityWorkerTimeout` and `MaxHeartbeatThrottleInterval` to `MaxNonLocalActivityHeartbeatThrottleInterval` and such if we're concerned.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: You would be much shorter processing times for remote codecs if you batched. Should we batch instead?

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: I have a lot of tests and things to verify that I don't have issues for

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: Agreed on what they should be, there was never any argument there. I am asking about what to do about Python knowing it already exists unencoded today. There are compatibility concerns. I will fix in .NET.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:122]: I don't need to make that assumption I don't think unless a microbenchmark proves otherwise.

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (bergundy) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: If you prefer false for default I would avoid the double negative.

I think I called non-local, remote activities a couple of years ago in the TS SDK but I haven't seen that concept anywhere in our docs.

`LocalActivityWorkerOnly` works for me or `DisableActivityPolling` (even though the latter doesn't account for eager activities), @Sushisource had issues with the local only wording.

Comment (cretz) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: :+1: We use the word "remote" in Rust and Python too. If @Sushisource doesn't have any objection, I'll switch to `LocalActivityWorkerOnly` even though a tad confusing because it at least matches Go and Java.

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowWorker.cs:122]: Agree it doesn't matter for perf reasons, this was mostly for my understanding of the guarantees.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: I meant for Python

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:195]: Done

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: As long as you're tracking this...

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: I agree that batching is better approach than parallelizing for both remote and non remote codecs.

We should just make sure that a single batch correlates to a single "context" (workflow / activity) to simplify the user API for when they need their codecs to be context aware.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:55]: > Headers should be encoded, they may contain sensitive information.
> I meant for Python

We know that headers should be encoded. The question is what to do for Python when users are already using unencoded headers today. The question is about compatibility not about whether they should be encoded. I can take this out of this PR and we can discuss what we should do in Python since the question is not coming across here.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: For performance reasons, should we update SDKs to support batching? Should we update .NET to support batching? Or should we accept that performance benefit is negligible compared to separate requests.

Comment (bergundy) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: I don't know how negligible it is, I think there are definitely cases where it would matter.
In any case I don't think we want to take this on for all SDKs right now.
In the meantime, batching could easily be built on top of concurrency user side by waiting a tick before submitting any accumulated requests (similarly to how this is done in [dataloader](https://www.npmjs.com/package/dataloader)) .

I wouldn't necessarily change the API and wouldn't block this PR, I would just take this into consideration.
In TypeScript we try to run codecs concurrently as much as possible (but I admit that there's room for improvement).

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:194]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowInstance.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:482]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:28]: Fixed in next PR to not complete the remove-job task until the hung workflow task is complete

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:90]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:146]: Fixed in next PR

File: src/Temporalio/Worker/TemporalWorkerOptions.cs
Comment (cretz) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:132]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:39]: While attempting to parallelize this, two other reasons came up to potentially not do it: 1) this will aggregate exceptions, so you could throw 50 exceptions from here for 50 commands, and 2) This can explode remote/IO call count filling up the thread pool. Technically they can prevent this by debouncing, but it is not as clear.

If you have a codec that does CPU bound AES let's say, and I call this concurrently for all 50 commands, .NET will try to use as much of the thread pool as it has available. Scatter-gather makes less sense for codecs for potentially IO/CPU-bound tasks in multithreaded languages.

As the implementation sits right now, we're not gonna use more than a single thread for a single workflow task, which is nice and predictable.

File: src/Temporalio/Worker/WorkflowCodecHelper.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowCodecHelper.cs:51]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowInstanceDetails.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstanceDetails.cs:29]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowWorker.cs
Comment (cretz) [reply src/Temporalio/Worker/WorkflowWorker.cs:217]: Fixed in next PR

**Pull request merged**
